File: e-dashboard-ui/src/ClientApp/src/views/Diagram/DiagramLib/Shapes/PortShape.ts
import { IPoint } from './interfaces';
import { Style } from './Style';
import { IStyle, IPortShape } from './interfaces';
import { Point } from './Shapes';
import { v4 as uuidv4 } from 'uuid';

export class PortShape implements IPortShape {
    public $type: string;
    public id: string;
    public name: string;
    public offset: IPoint;
    public location: IPoint;
    public isConnected: boolean;
    public handle: number ;
    public selectionZoneWidth: number ;
    public connectShapeId: string | null ;
    public style: IStyle ;

    constructor(name: string, offset: IPoint) {
        this.$type = "PortShape";
        this.id = uuidv4();
        this.name = name;
        this.offset = offset;
        this.location = new Point(0,0);
        this.isConnected = false;
        this.handle = 0;
        this.selectionZoneWidth = 3;
        this.connectShapeId = null;
        this.style = new Style();
        
        this.style.opacity = 1;
        this.style.fillStyle = "#FFFFFF00";
        this.style.strokeStyle = "#FF0000";
        this.style.strokeWidth = 1;
    }

    public draw(ctx: CanvasRenderingContext2D) {
        if (!this.isConnected) {
            var d = this.selectionZoneWidth;
            ctx.globalAlpha = this.style.opacity;
            ctx.lineWidth = this.style.strokeWidth;
            ctx.strokeStyle = this.style.strokeStyle;
            ctx.fillStyle = this.style.fillStyle;

            ctx.strokeRect(this.location.x + this.offset.x - d, this.location.y + this.offset.y - d, 2 * d, 2 * d);
        }
    }
    public move(to: IPoint) {
        this.location = to;
    }

    public getCursorType(mousePoint: IPoint): string {
        return "pointer";
    }
    public inPortZone(point: IPoint): boolean {
        let x = this.location.x + this.offset.x;
        let y = this.location.y + this.offset.y;
        let loc = new Point(x,y)
        return loc.distance(point) < this.selectionZoneWidth;
    }

    getLocation(): IPoint {
        let x = this.location.x + this.offset.x;
        let y = this.location.y + this.offset.y;
        return new Point(x, y)
    }
}

File: e-dashboard-api/src/Domain/Diagram/Shapes/PortShape.cs

using System.Text.Json.Serialization;

namespace EnergyDashboard.Domain.Diagram.Shapes;

[JsonDerivedType(typeof(PortShape), nameof(PortShape))]
public class PortShape
{
    public PortShape()
    {
        this.Id = Guid.NewGuid().ToString();
        this.Location = new Point(0, 0);
        this.IsConnected = false;
        this.Handle = 0;
        this.SelectionZoneWidth = 3;
        this.ConnectShapeId = string.Empty;
        this.Style = new Style();
        this.Style.Opacity = 1;
        this.Style.FillStyle = "#FFFFFF00";
        this.Style.StrokeStyle = "#FF0000";
        this.Style.StrokeWidth = 1;
    }
    public PortShape(string name, Point offset):this()
    {
        this.Name = name;
        this.Offset = offset;
    }

    public string Id { get; set; }
    public string Name { get; set; }
    public Point Offset { get; set; }
    public Point Location { get; set; }
    public bool IsConnected { get; set; }
    public int Handle { get; set; }
    public int SelectionZoneWidth { get; set; }
    public string ConnectShapeId { get; set; }
    public Style Style { get; set; }
}
